spec.html

```diff
@@ -3045,15 +3045,14 @@ <h1>Await</h1>
         <p>Algorithm steps that say</p>
 
         <emu-alg>
-          1. Let _completion_ be Await(_promise_).
+          1. Let _completion_ be Await(_value_).
         </emu-alg>
 
         <p>mean the same thing as:</p>
 
         <emu-alg>
           1. Let _asyncContext_ be the running execution context.
-          1. Let _promiseCapability_ be ! NewPromiseCapability(%Promise%).
-          1. Perform ! Call(_promiseCapability_.[[Resolve]], *undefined*, &laquo; _promise_ &raquo;).
+          1. Let _promise_ be ? PromiseResolve(&laquo; _value_ &raquo;).
```

Comment:
`PromiseResolve` currently takes a constructor in the first argument position. I suppose that we'll need to provide it with `%Promise%`.

Reply:
What are the chances we could tolerate any `Promise` subclass here? In other words, pass _value_`.constructor` as the first argument to `PromiseResolve` if _value_ `instanceof %Promise%` (pardon my hand waving; there may be a better spec notation for this), thereby giving _value_ a chance to be returned as-is by `PromiseResolve`.

I realize the `async` function can only return instances of the original, native `Promise` constructor, but it seems like a good thing for `await` to treat `Promise` subclasses the same way it treats ordinary `Promise`s. That is, unless part of the deal with subclassing `Promise` is that you're guaranteed your `.prototype.then` method will be called by `await`?

I don't have strong feelings either way, but it seems like a discussion worth having.

Reply:
A Promise subclass would (presumably) fail the `SameValue` test on 2.b of `PromiseResolve`, and we would then fallback to creating a new promise (exactly as we currently do).

Reply:
I believe it was an intentional design decision, discussed in committee, not to support subclassed Promises in `await`.

Reply:
(Specifically, using .constructor here would mean that any object coercible value in the language would pass the test in PromiseResolve, which would break the semantics)

Reply:
So it's exactly as if `await` calls `Promise.resolve`, which [passes `this` as the constructor argument to `PromiseResolve`](https://tc39.github.io/ecma262/#sec-promise.resolve). Although `this` could be a subclass of `Promise` in arbitrary user code, it would always be `%Promise%` in an `async` function.

Reply:
> (Specifically, using .constructor here would mean that any object coercible value in the language would pass the test in PromiseResolve, which would break the semantics)

That's why we would have to perform an external check that `.constructor` is a `Promise` subclass before calling `PromiseResolve(_value_.constructor, _value_)`, or perhaps modify `PromiseResolve` to enforce the subclass relationship.

How do folks feel about introducing a single-argument `PromiseResolve` variant that does the right thing (whatever we decide that is), so that this code (i.e. `Let _promise_ be ? PromiseResolve(&laquo; _value_ &raquo;)`) works as written?

Reply:
I would probably prefer the explicit constructor argument to `PromiseResolve` for now. And I believe it should be `%Promise%` in any case: subclasses should go through the slower but semantically more flexible path of calling their overridden "then" method from a new tick.

Reply:
I also think it's useful to keep the explicit `%Promise%`.

Reply:
Thanks for the discussion!
Added `%Promise%` as a parameter in a follow-up commit.